presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java

```diff
@@ -138,36 +142,38 @@ public void matches(PlanMatchPattern pattern)
 
         inTransaction(session -> {
             Map<PlanNodeId, PlanNodeCost> planNodeCosts = costCalculator.calculateCostForPlan(session, types, actual);
-            assertPlan(session, metadata, costCalculator, new Plan(actual, types, planNodeCosts), ruleApplication.lookup, pattern);
+            assertPlan(session, metadata, costCalculator, new Plan(actual, types, planNodeCosts), result.lookup, pattern);
             return null;
         });
     }
 
-    private RuleApplication applyRule()
+    private OptimizationResult optimize()
     {
-        SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
-        Memo memo = new Memo(idAllocator, plan);
-        Lookup lookup = Lookup.from(planNode -> Stream.of(memo.resolve(planNode)));
-
-        PlanNode memoRoot = memo.getNode(memo.getRootGroup());
-
-        return inTransaction(session -> applyRule(rule, memoRoot, ruleContext(symbolAllocator, lookup, session)));
-    }
-
-    private static <T> RuleApplication applyRule(Rule<T> rule, PlanNode planNode, Rule.Context context)
-    {
-        PlanNodeMatcher matcher = new PlanNodeMatcher(context.getLookup());
-        Match<T> match = matcher.match(rule.getPattern(), planNode);
-
-        Optional<PlanNode> result;
-        if (!rule.isEnabled(context.getSession()) || match.isEmpty()) {
-            result = Optional.empty();
-        }
-        else {
-            result = rule.apply(match.value(), match.captures(), context);
-        }
+        return inTransaction(session -> {
+            List<RuleApplicationListener.RuleApplication> ruleApplications = new ArrayList<>();
+            IterativeOptimizer optimizer = new IterativeOptimizer(
+                    new StatsRecorder(),
+                    ruleApplications::add,
+                    ImmutableSet.of(rule));
+
+            SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
+            Memo memo = new Memo(idAllocator, plan);
+            Lookup lookup = Lookup.from(planNode -> Stream.of(memo.resolve(planNode)));
+            optimizer.optimize(memo, session, symbolAllocator, idAllocator);
```

Comment:
memo is internal state of the Optimizer. Should not be provided from outside and there should be no need for that here.

Reply:
Memo is not an internal state of optimizer. It was already used in the same code here.

Memo is a structure the optimizer works on. You could apply different optimizers to reorganize `Memo` (e.g: `GreedyOptimizer`, `ExhaustiveOptimizer`, etc...). 

In fact I think there might be some phases in equivalence optimization, where we will apply different rules in phases (possibly even by different optimizers), e.g:
```
new ExhaustiveOptimizer(PHASE_1_RULES).optimize(memo); 
...
new ExhaustiveOptimizer(PHASE_2_RULES).optimize(memo); 
``` 

I think `Memo` will actually become an interface in the near future.